### ISSUE TITLE
Reload node definitions from a job workflow

### ISSUE THREAD

**Issue by gschueler** [rundeck/rundeck#78]
If I'm using a job to spin up new instances in EC2, it would be nice if the workflow could be something like:
- Spin up new instances.
- Reload node definitions from my resource provider (Chef).
- Use those new nodes for subsequent steps in the job.

[original LH ticket](http://rundeck.lighthouseapp.com/projects/59277/tickets/328)

Original Creator: Bruce Mitchener

**Comment by danielAparicio** [comment]
Any news about this Issue??

**Comment by scaffster** [comment]
We have hit the same issue, the ec2 plugin pulls the nodes in at job creation, but requires a refresh to pull in any node changes at job execution run-time.

**Comment by jinesh-choksi** [comment]
We also hit this issue with the ec2 plugin. In our case, we were destroying/creating nodes frequently and would get a "Failed dispatching to node OUR_NODE_NAME: java.lang.IllegalArgumentException: Null hostname value" error when executing a job which had a _valid_ node filter.

The ec2 plugin maintains a resource cache in a folder called ${framework.var.dir}/resourceModelSourceCache/YOUR_PROJECT_NAME.
(The value of framework.var.dir can be found in rundeck's framework.properties file.)

In my case the cache files were called: 2.source.xml & cache.xml.

I found that when updating an EC2 node's tags via AWS console, the rundeck ec2 plugin's cache was not getting updated with those changes until I browsed to the project's Nodes page. Simply browsing to the project's Nodes page was forcing a cache refresh.

I also found that the following curl command had the same effect as browsing to the project's Nodes page:

`curl --silent --show-error --location --header "X-RunDeck-Auth-Token: ###YOUR_AUTH_TOKEN###" "https://rundeck.example.com/api/17/resources?project=YOUR_PROJECT_NAME&format=xml"`

(You can generate an auth token by clicking on your profile in Rundeck and clicking the "Generate New Token" link next to "API Tokens".)

So as a workaround we created a meta job

- which executes locally on the rundeck server
- which calls the above curl statement to refresh the resource cache as its first workflow step
- and then its next workflow step is to call the actual job which previously failed to run due to a stale resource cache

Its not clear whether the resource node cache is not being updated due to a bug in the ec2 plugin or whether its an issue with rundeck's resource caching framework.

**Comment by daugustus** [comment]
Using the Curl script was a good workaround. IMHO, you either have to trigger the node cache to refresh or wait for the nodeCache.delay to expire (in the case of the Ec2 node resource).  Rundeck would have to have knowledge, via the plugin, that a new node was created and therefore the cache needs to expire. 

My preference is to run the curl script. 

Perhaps another option would be to have resource plugin refresh via an API call

**Comment by mattmauriello** [comment]
@jinesh-choksi 
I'm doing some work myself with rundeck in AWS. I'm hoping to put some stuff together to use SSM as a node executor and file copier, but was hoping you could clarify your experience here. Did the cache NEVER update, or did it update after some time, which was too long for use case forcing you to do the manual refresh?

**Comment by daugustus** [comment]
There is a cache timeout setting for the ec2 plugin default of 30 seconds. Waiting for 30 seconds (or more) didn't seem to help, once the job started.  

It would seem that the data had to exist BEFORE the job started, in order to be available DURING the job. In my case, I am creating nodes DURING the job therefore the contents of the cache is irrelevant but also does not expire. That is my guess, anyway. The Curl script worked so I moved on to solving other problems. ;)

**Comment by jinesh-choksi** [comment]
We found that the ec2 node cache does have an effect on our meta job. If the cache had active content then no matter how many curl requests we made, rundeck just got the active state from the cache instead of querying the ec2 api. Of course to work around this we set the ec2 node cache refresh period to a low value. I think my manager found that setting it to 0 didn't disable the cache and he was considering amending the plugin code to fix that. I will find out what he did on Monday when I return to work.

**Comment by mattmauriello** [comment]
well, i was doing some digging, on a fresh install of 2.7.1, watching the cache file and checking he job runs. I had a job run every minute, with a node filter set based on a tag on an ec2 instance, with a refresh interval on the EC2 nodes plugin set to 600. i changed the tags around a few times through a test. what i saw was that the cace files were updated EVERY run, and the job ran against the node that matched, based on the actual state in AWS.

So for now, i guess im dropping this until i see it behave differently.

**Comment by burdandrei** [comment]
@jinesh-choksi we're doing something similar, before the actual job trigger, we are triggering that sends the hostame command to all the nodes. It fails because of dead nodes, but then we receive actual node list for the real job

**Comment by jbguerraz** [comment]
We've got a (probably) similar issue.
We're deploying rundeck using ansible, and right after the deployment task we setup a project, and a job and then execute the job (still using ansible => straight in a row). The job fails to execute because it can't get nodes that match the node filter (which is ".*"). If then we open the user interface and load the "nodes" page of the given project then the job will execute successfully : so, it looks like rundeck won't ask the resources source plugin for the list of nodes at job execution and will rely only on its cached list of nodes (if there is, if not, it'll fails).
Disabling the node cache (project.nodeCache.enabled=false) make it working properly.

**Comment by gschueler** [comment]
Hi all,

1. the ec2 plugin has its own asynchronous cache behavior, but it is also on-demand, rather than actively refreshing periodically, so it doesn't refresh unless asked.  It won't be asked by rundeck until rundeck's internal cached nodes are out of date.  so even if the rundeck cache is out of date, and it refreshes the ec2 plugin, that plugin returns its cached data while it reloads from AWS in the background, so not until the next time the Rundeck cache expires (and asynchronously asks the plugin for data) will the data be refreshed with the now long out of date nodes. But setting the refresh delay to `0` in the plugin still causes the out of date problem because it there are two levels of async loading happening. There needs to be a way of loading synchronously in the ec2 plugin.  I have filed an issue to have a way to force synchronous loading in the plugin: https://github.com/rundeck-plugins/rundeck-ec2-nodes-plugin/issues/61

2. for @jbguerraz you can keep the node cache on, but force the *first load* of nodes data to be done synchronously using the `rundeck.nodeService.nodeCache.firstLoadAsynch=false` configuration value. Maybe this should be the default, but we set it to `true` because if you have a lot of projects with a lot of nodes, the app startup can take an extremely long time.  Maybe a middle-ground solution is to load synchronously the first time with a max timeout, and after that timeout do the rest in the background.

3. the original issue here is refreshing the nodes in the middle of a workflow, and that is a feature that still needs to be added.

hth

**Comment by EugenMayer** [comment]
just in case, that curl request uses outdated/deprecated api, you should use `/api/21/project/<project>/resource` GET

**Comment by cdstg** [comment]
yes I have the same problem it would be nice to have that API call at the moment I called 
/api/21/project/<project>/resources GET

BUT even then I have to wait 30 seconds or so before running my job as it seems that it does not update the cache straight away for some reason with the dynamic inventory 

I m not sure if there is some setting I have not got right in the node definition 

It does work but waiting 30 seconds ( or 45 to be sure ) is a pain

**Comment by angrybb** [comment]
This is still relevant in Rundeck 3.0.23-20190619

**Comment by gschueler** [comment]
a step plugin was added in 2.8 that can perform this action https://github.com/rundeck/rundeck/pull/2380
